sd-validation/schema.js

```diff
@@ -0,0 +1,144 @@
+/**
+ * @license Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const walkObject = require('./helpers/walkObject.js');
+// @ts-ignore
+const schemaStructure = require('./assets/schema-tree.json');
+const TYPE_KEYWORD = '@type';
+const SCHEMA_ORG_URL_REGEX = /https?:\/\/schema\.org\//;
+
+/**
+ * @param {string} uri
+ * @returns {string}
+ */
+function cleanName(uri) {
+  return uri.replace(SCHEMA_ORG_URL_REGEX, '');
+}
+
+/**
+ * @param {string} type
+ * @returns {Array<string>}
+ */
+function getPropsForType(type) {
+  const cleanType = cleanName(type);
+  const props = schemaStructure.properties
+    .filter((/** @type {{name: string, parent: Array<string>}} */ prop) =>
+      prop.parent.includes(cleanType))
+    .map((/** @type {{name: string, parent: Array<string>}} */ prop) => prop.name);
+  const foundType = findType(type);
+  if (!foundType) throw new Error(`Unable to get props for missing type "${type}"`);
+  const parentTypes = foundType.parent;
+
+  return parentTypes.reduce((allProps, type) => allProps.concat(getPropsForType(type)), props);
+}
+
+/**
+ * @param {string} type
+ * @returns {{name: string, parent: Array<string>}|undefined}
+ */
+function findType(type) {
+  const cleanType = cleanName(type);
+
+  return schemaStructure.types
+    .find((/** @type {{name: string}} */ typeObj) => typeObj.name === cleanType);
+}
+
+/**
+ * @param {string} type
+ * @returns {boolean}
+ */
+function isKnownType(type) {
+  return findType(type) !== undefined;
+}
+
+/**
+ * Validates keys of given object based on it's type(s). Returns an array of error messages.
+ *
+ * @param {String|Array<String>} typeOrTypes
+ * @param {Array<String>} keys
+ * @returns {Array<String>}
+ */
+function validateObjectKeys(typeOrTypes, keys) {
+  /** @type {Array<string>} */
+  const errors = [];
+  /** @type {Array<string>} */
+  const safelist = [];
+
+  let types = [];
+
+  if (typeof typeOrTypes === 'string') {
+    types.push(typeOrTypes);
+  } else if (Array.isArray(typeOrTypes)) {
+    types = typeOrTypes;
+  } else {
+    return ['Unknown value type'];
+  }
+
+  const unknownTypes = types.filter(t => !isKnownType(t));
+
+  unknownTypes
+    .forEach(type => {
+      if (typeof type !== 'string' || SCHEMA_ORG_URL_REGEX.test(type)) {
```

Comment:
if it's an unknown type string but it doesn't pass `SCHEMA_ORG_URL_REGEX.test()`, what does that mean? It seems like it would throw when it gets to `getPropsForType`...shouldn't it just push an error message now?

Reply:
no, `getPropsForType` won't be called if there are any unknown types. if all the unknown types are not schema types then we return success via empty errors array

Reply:
restructued to make this clearer